rdf-abac-fuseki/src/main/java/io/telicent/jena/abac/fuseki/server/UserInfoEnrichmentFilter.java

```diff
@@ -0,0 +1,457 @@
+     * @throws ServletException if the chain throws
+     */
+    @Override
+    public void doFilter(ServletRequest req, ServletResponse resp, FilterChain chain)
+            throws IOException, ServletException {
+        HttpServletRequest httpReq = (HttpServletRequest) req;
+        String authz = httpReq.getHeader(HttpNames.hAuthorization);
+
+        if (authz != null && authz.toLowerCase(Locale.ROOT).startsWith("bearer ")) {
+            String token = authz.substring(7).trim();
+            if (!token.isBlank()) {
+                String username = AttributesStoreAuthServer.getCachedUsername(token);
+                if (username != null) {
+                    req.setAttribute(ATTR_ABAC_USERNAME, username);
+                } else {
+                    CachedUser cu = fetchUserInfo(token);
+                    if (cu != null && cu.attributes != null && cu.username != null) {
+                        req.setAttribute(ATTR_ABAC_USERNAME, cu.username);
+                        // Populate cross-request caches for downstream usage.
+                        AttributesStoreAuthServer.addIdAndUserName(token, cu.username);
+                        AttributesStoreAuthServer.addUserNameAndAttributes(cu.username, cu.attributes);
+                    }
+                    if (cu == null && legacyFilter != null) {
+                        this.legacyFilter.doFilter(req, resp, chain);
+                        return;
+                    }
+                }
+            }
+        }
+        chain.doFilter(req, resp);
+    }
+
+    /**
+     * Immutable container for userinfo results.
+     *
+     * @param username   resolved username
+     * @param attributes flattened/normalized attribute set
+     */
+    private record CachedUser(String username, AttributeValueSet attributes) {
+    }
+
+    /**
+     * Retrieve and parse user information from the configured {@code /userinfo} endpoint.
+     *
+     * <p>Expected payload examples:</p>
+     * <ul>
+     *   <li>{@code preferred_username}/{@code username}/{@code name}/{@code sub} for the username, and</li>
+     *   <li>either {@code attributes.abac_attributes} as an array of {@code "k=v"} strings, or a
+     *       structured {@code attributes} object which is flattened into {@code "k=v"} pairs.</li>
+     * </ul>
+     *
+     * @param token the Bearer token value (without the {@code "Bearer "} prefix)
+     * @return a {@link CachedUser} if successful; otherwise {@code null} (on non-2xx or any exception)
+     */
+    private CachedUser fetchUserInfo(String token) {
+        try {
+            HttpRequest request = HttpRequest.newBuilder(URI.create(userInfoUrl))
+                    .timeout(Duration.ofSeconds(5))
+                    .header(HttpNames.hAuthorization, "Bearer " + token)
+                    .GET().build();
+
+            HttpResponse<String> response = http.send(request, HttpResponse.BodyHandlers.ofString());
+            if (response.statusCode() / 100 != 2) {
```

Comment:
```suggestion
            if (response.statusCode() >= 200 && response.statusCode() >= 300) {
```

Is more long winded but probably clearer to read for our future selves